contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+
+Prebound PVCs and PVs will still immediately be bound by the PV controller.
+
+One important point to note is that for the alpha phase, manual recovery is
+required in following error conditions:
+* A Pod has multiple PVCs, and only a subset of them successfully bind.
+* The scheduler chose a PV and prebound it, but the PVC could not be bound.
+
+In both of these scenarios, the primary cause for these errors is if a user
+prebinds other PVCs to a PV that the scheduler chose.  Some workarounds to
+avoid these error conditions are to:
+* Prebind the PV instead.
+* Separate out volumes that the user prebinds from the volumes that are
+available for the system to choose from by StorageClass.
+
+In addition, this new design will cause functional regression for the
+following use cases:
+* While a Pod is running, a user deletes the PVC and PV, and then recreates them.
+    * With the current behavior, the PVC will rebind to the same or different PV.
+    * With this new behavior, the PVC will not rebind because binding would only
+be triggered when the Pod has to go through scheduling.  This could cause
+problems with Pod volume cleanup because the PVC will no longer be bound.
+    * But deleting the PVC and PV (especially both) while a Pod is still using it
+is also dangerous and should be prevented in the first place.  There could be
+serious problems if you recreate the PVC and it rebinds to a different PV.  If
+the PV had a reclaim policy of Delete, then deleting the PVC could cause the
+underlying disk to get deleted while the Pod is still using it.
+
+
+#### PV Controller Changes
+When the feature gate is enabled, the PV controller needs to skip binding and
+provisioning all unbound PVCs that don’t have prebound PVs, and let it come
+through the scheduler path.  This is the block in `syncUnboundClaim` where
+`claim.Spec.VolumeName == “”` and there is no PV that is prebound to the PVC.
+
+No other state machine changes are required.  The PV controller continues to
+handle the remaining scenarios without any change.
+
+The methods to find matching PVs for a claim, prebind PVs, and to invoke and
+handle dynamic provisioning need to be refactored for use by the new scheduler
+functions.
+
+#### Scheduler Changes
+
+##### Predicate
+A new predicate function checks all of a Pod's unbound PVCs can be satisfied
+by existing PVs or dynamically provisioned PVs that are
+topologically-constrained to the Node.
+```
+MatchUnboundPVCs(pod *v1.Pod, node *v1.Node) (canBeBound bool, err error)
+```
+1. If all the Pod’s PVCs are bound, return true.
+2. Otherwise try to find matching PVs for all of the unbound PVCs in order of
+decreasing requested capacity.
+3. Walk through all the PVs.
+4. Find best matching PV for the PVC where PV topology is satisfied by the Node.
+5. Temporarily cache this PV in the PVC object, keyed by Node, for fast
+processing later in the priority and bind functions.
+6. Return true if all PVCs are matched.
```

Comment:
What happens if matched PVs are deleted or PVC is resized since scheduler started matching? I am just trying to think, there appears to be a delay between when PVs/node is matched and when binding happens. Does the pod go through `MatchUnbounPVCs` again if something changed since match began?